### PR TITLE
Replace failwith with Code_error or Jsonrpc.Response.Error

### PR DESCRIPTION
**Pull request by tmattio**

### CODE REVIEW

File: ocaml-lsp-server/src/inference.ml [hunk 1]
```diff
@@ -2,7 +2,9 @@ open Import
 
 let infer_intf_for_impl doc =
   match Document.kind doc with
-  | Intf -> failwith "the provided document is not an implementation."
+  | Intf ->
+    Code_error.raise
+      "expected an implementation document, got an interface instead" []
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/inference.ml:7]: Is it possible for the user to trigger this error?

File: ocaml-lsp-server/src/document.ml [hunk 2]
```diff
@@ -13,7 +13,9 @@ module Kind = struct
     | ".mli"
     | ".rei" ->
       Intf
-    | ext -> failwith ("Unknown extension " ^ ext)
+    | ext ->
+      Code_error.raise "unsupported file extension"
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/document.ml:17]: Is it possible for the user to trigger this error?

**Review by rgrinberg — APPROVED** [review]
Looks good. On the account that it's not possible for the user to trigger the code errors.

File: ocaml-lsp-server/src/inference.ml (hunk 1 above)
Comment (tmattio) [reply ocaml-lsp-server/src/inference.ml:7]: The input argument is a `Document.t`, and from what I can see the only way to get that is to get the doc from the store. So a user should not be able to trigger this unless there's a bug.

File: ocaml-lsp-server/src/document.ml (hunk 2 above)
Comment (tmattio) [reply ocaml-lsp-server/src/document.ml:17]: The input argument is a `Document.t`, and from what I can see the only way to get that is to get the doc from the store. So a user should not be able to trigger this unless there's a bug.

Comment (tmattio) [reply ocaml-lsp-server/src/document.ml:17]: That's not true, I wasn't looking at the right place.

**Review by rgrinberg — APPROVED** [review]

File: ocaml-lsp-server/src/document.ml [hunk 3]
```diff
@@ -13,7 +13,11 @@ module Kind = struct
     | ".mli"
     | ".rei" ->
       Intf
-    | ext -> failwith ("Unknown extension " ^ ext)
+    | ext ->
+      Jsonrpc.Response.Error.raise
+        (Jsonrpc.Response.Error.make ~code:InvalidRequest
+           ~message:(Printf.sprintf "unsupported file extension %S" ext)
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/document.ml:19]: Technically `ext` should occur inside the error's payload

**Pull request merged**